src/track/globaltrackcache.cpp

```diff
@@ -300,296 +412,341 @@ TrackPointer GlobalTrackCache::lookupInternal(
+TrackRef GlobalTrackCache::initTrackId(
+        const TrackPointer& strongPtr,
+        TrackRef trackRef,
         TrackId trackId) {
+    DEBUG_ASSERT(strongPtr);
+    DEBUG_ASSERT(!trackRef.getId().isValid());
     DEBUG_ASSERT(trackId.isValid());
-    if (trackRef.getId() != trackId) {
-        DEBUG_ASSERT(m_tracksById.end() == m_tracksById.find(trackId));
-        TrackRef trackRefWithId(trackRef, trackId);
-        Item item(trackRefWithId, pTrack);
-        m_tracksById.insert(
-                item.ref.getId(),
-                item);
-        m_tracksByCanonicalLocation.insert(
-                item.ref.getCanonicalLocation(),
-                item);
-        return trackRefWithId;
-    }
-    return trackRef;
-}
-
-void GlobalTrackCache::evict(
-        Track* pTrack) {
-    DEBUG_ASSERT(pTrack != nullptr);
-    const auto trackRef = TrackRef::fromFileInfo(
-            pTrack->m_fileInfo,
-            pTrack->m_record.getId());
-    GlobalTrackCacheLocker cacheLocker;
-    Track* pEvictedTrack = evictInternal(&cacheLocker, trackRef);
-    // The cache might have been unlocked during the callback!
-    DEBUG_ASSERT((nullptr == pEvictedTrack) || (pEvictedTrack == pTrack));
+    DEBUG_ASSERT(m_tracksByCanonicalLocation.find(
+            trackRef.getCanonicalLocation()) != m_tracksByCanonicalLocation.end());
+    TrackRef trackRefWithId(trackRef, trackId);
+
+    // Insert item by id
+    DEBUG_ASSERT(m_tracksById.find(trackId) == m_tracksById.end());
+    m_tracksById.insert(std::make_pair(
+            trackId,
+            strongPtr.get()));
+
+    strongPtr->initId(trackId);
+    DEBUG_ASSERT(createTrackRef(*strongPtr) == trackRefWithId);
+
+    return trackRefWithId;
 }
 
-GlobalTrackCache::Item GlobalTrackCache::purgeInternal(
-        const TrackRef& trackRef) {
-    if (kLogger.debugEnabled()) {
-        kLogger.debug()
-                << "Purging track"
-                << trackRef;
-    }
+void GlobalTrackCache::afterEvicted(
+        GlobalTrackCacheLocker* pCacheLocker,
+        Track* pEvictedTrack) {
+    DEBUG_ASSERT(pEvictedTrack);
+
+    // It can produce dangerous signal loops if the track is still
+    // sending signals while being saved! All references to this
+    // track have been dropped at this point, so there is no need
+    // to send any signals.
+    // See: https://bugs.launchpad.net/mixxx/+bug/136578
+    // NOTE(uklotzde, 2018-02-03): Simply disconnecting all receivers
+    // doesn't seem to work reliably. Emitting the clean() signal from
+    // a track that is about to deleted may cause access violations!!
+    pEvictedTrack->blockSignals(true);
+
+    // Keep the cache locked while evicting the track object!
+    // The callback is given the chance to unlock the cache
+    // after all operations that rely on managed track ownership
+    // have been done, e.g. exporting track metadata into a file.
+    m_pEvictor->afterEvictedTrackFromCache(
+            pCacheLocker,
```

Comment:
We cannot use the qt queue trick, because this introduces a deadlock risk, because we pass a locked mutex onto the queue.